// angular2/ts/src/alt_router/segments.js
System.register(['angular2/src/facade/collection', 'angular2/src/facade/lang'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var collection_1, lang_1;
    var Tree, UrlSegment, RouteSegment;
    function equalSegments(a, b) {
        if (lang_1.isBlank(a) && !lang_1.isBlank(b))
            return false;
        if (!lang_1.isBlank(a) && lang_1.isBlank(b))
            return false;
        return a._type === b._type && collection_1.StringMapWrapper.equals(a._parameters, b._parameters);
    }
    exports_1("equalSegments", equalSegments);
    function routeSegmentComponentFactory(a) {
        return a._componentFactory;
    }
    exports_1("routeSegmentComponentFactory", routeSegmentComponentFactory);
    return {
        setters:[
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            Tree = (function () {
                function Tree(_nodes) {
                    this._nodes = _nodes;
                }
                Object.defineProperty(Tree.prototype, "root", {
                    get: function () { return this._nodes[0]; },
                    enumerable: true,
                    configurable: true
                });
                Tree.prototype.parent = function (t) {
                    var index = this._nodes.indexOf(t);
                    return index > 0 ? this._nodes[index - 1] : null;
                };
                Tree.prototype.children = function (t) {
                    var index = this._nodes.indexOf(t);
                    return index > -1 && index < this._nodes.length - 1 ? [this._nodes[index + 1]] : [];
                };
                Tree.prototype.firstChild = function (t) {
                    var index = this._nodes.indexOf(t);
                    return index > -1 && index < this._nodes.length - 1 ? this._nodes[index + 1] : null;
                };
                Tree.prototype.pathToRoot = function (t) {
                    var index = this._nodes.indexOf(t);
                    return index > -1 ? this._nodes.slice(0, index + 1) : null;
                };
                return Tree;
            }());
            exports_1("Tree", Tree);
            UrlSegment = (function () {
                function UrlSegment(segment, parameters, outlet) {
                    this.segment = segment;
                    this.parameters = parameters;
                    this.outlet = outlet;
                }
                return UrlSegment;
            }());
            exports_1("UrlSegment", UrlSegment);
            RouteSegment = (function () {
                function RouteSegment(urlSegments, parameters, outlet, type, componentFactory) {
                    this.urlSegments = urlSegments;
                    this.outlet = outlet;
                    this._type = type;
                    this._componentFactory = componentFactory;
                    this._parameters = parameters;
                }
                RouteSegment.prototype.getParam = function (param) { return this._parameters[param]; };
                Object.defineProperty(RouteSegment.prototype, "type", {
                    get: function () { return this._type; },
                    enumerable: true,
                    configurable: true
                });
                return RouteSegment;
            }());
            exports_1("RouteSegment", RouteSegment);
        }
    }
});
