// angular2/ts/src/core/linker/component_factory.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/exceptions', './view_utils'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __extends = (this && this.__extends) || function (d, b) {
        for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var lang_1, exceptions_1, view_utils_1;
    var ComponentRef, ComponentRef_, ComponentFactory;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (view_utils_1_1) {
                view_utils_1 = view_utils_1_1;
            }],
        execute: function() {
            /**
             * Represents an instance of a Component created via a {@link ComponentFactory}.
             *
             * `ComponentRef` provides access to the Component Instance as well other objects related to this
             * Component Instance and allows you to destroy the Component Instance via the {@link #destroy}
             * method.
             */
            ComponentRef = (function () {
                function ComponentRef() {
                }
                Object.defineProperty(ComponentRef.prototype, "location", {
                    /**
                     * Location of the Host Element of this Component Instance.
                     */
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ComponentRef.prototype, "injector", {
                    /**
                     * The injector on which the component instance exists.
                     */
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ComponentRef.prototype, "instance", {
                    /**
                     * The instance of the Component.
                     */
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                ;
                Object.defineProperty(ComponentRef.prototype, "hostView", {
                    /**
                     * The {@link ViewRef} of the Host View of this Component instance.
                     */
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                ;
                Object.defineProperty(ComponentRef.prototype, "changeDetectorRef", {
                    /**
                     * The {@link ChangeDetectorRef} of the Component instance.
                     */
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ComponentRef.prototype, "componentType", {
                    /**
                     * The component type.
                     */
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                return ComponentRef;
            }());
            exports_1("ComponentRef", ComponentRef);
            ComponentRef_ = (function (_super) {
                __extends(ComponentRef_, _super);
                function ComponentRef_(_hostElement, _componentType) {
                    _super.call(this);
                    this._hostElement = _hostElement;
                    this._componentType = _componentType;
                }
                Object.defineProperty(ComponentRef_.prototype, "location", {
                    get: function () { return this._hostElement.elementRef; },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ComponentRef_.prototype, "injector", {
                    get: function () { return this._hostElement.injector; },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ComponentRef_.prototype, "instance", {
                    get: function () { return this._hostElement.component; },
                    enumerable: true,
                    configurable: true
                });
                ;
                Object.defineProperty(ComponentRef_.prototype, "hostView", {
                    get: function () { return this._hostElement.parentView.ref; },
                    enumerable: true,
                    configurable: true
                });
                ;
                Object.defineProperty(ComponentRef_.prototype, "changeDetectorRef", {
                    get: function () { return this.hostView; },
                    enumerable: true,
                    configurable: true
                });
                ;
                Object.defineProperty(ComponentRef_.prototype, "componentType", {
                    get: function () { return this._componentType; },
                    enumerable: true,
                    configurable: true
                });
                ComponentRef_.prototype.destroy = function () { this._hostElement.parentView.destroy(); };
                ComponentRef_.prototype.onDestroy = function (callback) { this.hostView.onDestroy(callback); };
                return ComponentRef_;
            }(ComponentRef));
            exports_1("ComponentRef_", ComponentRef_);
            ComponentFactory = (function () {
                function ComponentFactory(selector, _viewFactory, _componentType) {
                    this.selector = selector;
                    this._viewFactory = _viewFactory;
                    this._componentType = _componentType;
                }
                Object.defineProperty(ComponentFactory.prototype, "componentType", {
                    get: function () { return this._componentType; },
                    enumerable: true,
                    configurable: true
                });
                /**
                 * Creates a new component.
                 */
                ComponentFactory.prototype.create = function (injector, projectableNodes, rootSelectorOrNode) {
                    if (projectableNodes === void 0) { projectableNodes = null; }
                    if (rootSelectorOrNode === void 0) { rootSelectorOrNode = null; }
                    var vu = injector.get(view_utils_1.ViewUtils);
                    if (lang_1.isBlank(projectableNodes)) {
                        projectableNodes = [];
                    }
                    // Note: Host views don't need a declarationAppElement!
                    var hostView = this._viewFactory(vu, injector, null);
                    var hostElement = hostView.create(projectableNodes, rootSelectorOrNode);
                    return new ComponentRef_(hostElement, this._componentType);
                };
                ComponentFactory = __decorate([
                    lang_1.CONST(), 
                    __metadata('design:paramtypes', [String, Function, lang_1.Type])
                ], ComponentFactory);
                return ComponentFactory;
            }());
            exports_1("ComponentFactory", ComponentFactory);
        }
    }
});
