// angular2/ts/src/core/linker/element.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/collection', 'angular2/src/facade/exceptions', './view_type', './element_ref', './view_container_ref'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, collection_1, exceptions_1, view_type_1, element_ref_1, view_container_ref_1;
    var AppElement;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (view_type_1_1) {
                view_type_1 = view_type_1_1;
            },
            function (element_ref_1_1) {
                element_ref_1 = element_ref_1_1;
            },
            function (view_container_ref_1_1) {
                view_container_ref_1 = view_container_ref_1_1;
            }],
        execute: function() {
            /**
             * An AppElement is created for elements that have a ViewContainerRef,
             * a nested component or a <template> element to keep data around
             * that is needed for later instantiations.
             */
            AppElement = (function () {
                function AppElement(index, parentIndex, parentView, nativeElement) {
                    this.index = index;
                    this.parentIndex = parentIndex;
                    this.parentView = parentView;
                    this.nativeElement = nativeElement;
                    this.nestedViews = null;
                    this.componentView = null;
                }
                Object.defineProperty(AppElement.prototype, "elementRef", {
                    get: function () { return new element_ref_1.ElementRef(this.nativeElement); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(AppElement.prototype, "vcRef", {
                    get: function () { return new view_container_ref_1.ViewContainerRef_(this); },
                    enumerable: true,
                    configurable: true
                });
                AppElement.prototype.initComponent = function (component, componentConstructorViewQueries, view) {
                    this.component = component;
                    this.componentConstructorViewQueries = componentConstructorViewQueries;
                    this.componentView = view;
                };
                Object.defineProperty(AppElement.prototype, "parentInjector", {
                    get: function () { return this.parentView.injector(this.parentIndex); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(AppElement.prototype, "injector", {
                    get: function () { return this.parentView.injector(this.index); },
                    enumerable: true,
                    configurable: true
                });
                AppElement.prototype.mapNestedViews = function (nestedViewClass, callback) {
                    var result = [];
                    if (lang_1.isPresent(this.nestedViews)) {
                        this.nestedViews.forEach(function (nestedView) {
                            if (nestedView.clazz === nestedViewClass) {
                                result.push(callback(nestedView));
                            }
                        });
                    }
                    return result;
                };
                AppElement.prototype.attachView = function (view, viewIndex) {
                    if (view.type === view_type_1.ViewType.COMPONENT) {
                        throw new exceptions_1.BaseException("Component views can't be moved!");
                    }
                    var nestedViews = this.nestedViews;
                    if (nestedViews == null) {
                        nestedViews = [];
                        this.nestedViews = nestedViews;
                    }
                    collection_1.ListWrapper.insert(nestedViews, viewIndex, view);
                    var refRenderNode;
                    if (viewIndex > 0) {
                        var prevView = nestedViews[viewIndex - 1];
                        refRenderNode = prevView.lastRootNode;
                    }
                    else {
                        refRenderNode = this.nativeElement;
                    }
                    if (lang_1.isPresent(refRenderNode)) {
                        view.renderer.attachViewAfter(refRenderNode, view.flatRootNodes);
                    }
                    view.addToContentChildren(this);
                };
                AppElement.prototype.detachView = function (viewIndex) {
                    var view = collection_1.ListWrapper.removeAt(this.nestedViews, viewIndex);
                    if (view.type === view_type_1.ViewType.COMPONENT) {
                        throw new exceptions_1.BaseException("Component views can't be moved!");
                    }
                    view.renderer.detachView(view.flatRootNodes);
                    view.removeFromContentChildren(this);
                    return view;
                };
                return AppElement;
            }());
            exports_1("AppElement", AppElement);
        }
    }
});
