// angular2/ts/src/core/render/api.js
System.register(['angular2/src/facade/exceptions'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var exceptions_1;
    var RenderComponentType, RenderDebugInfo, Renderer, RootRenderer;
    return {
        setters:[
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            }],
        execute: function() {
            RenderComponentType = (function () {
                function RenderComponentType(id, templateUrl, slotCount, encapsulation, styles) {
                    this.id = id;
                    this.templateUrl = templateUrl;
                    this.slotCount = slotCount;
                    this.encapsulation = encapsulation;
                    this.styles = styles;
                }
                return RenderComponentType;
            }());
            exports_1("RenderComponentType", RenderComponentType);
            RenderDebugInfo = (function () {
                function RenderDebugInfo() {
                }
                Object.defineProperty(RenderDebugInfo.prototype, "injector", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderDebugInfo.prototype, "component", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderDebugInfo.prototype, "providerTokens", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderDebugInfo.prototype, "locals", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderDebugInfo.prototype, "source", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                return RenderDebugInfo;
            }());
            exports_1("RenderDebugInfo", RenderDebugInfo);
            Renderer = (function () {
                function Renderer() {
                }
                return Renderer;
            }());
            exports_1("Renderer", Renderer);
            /**
             * Injectable service that provides a low-level interface for modifying the UI.
             *
             * Use this service to bypass Angular's templating and make custom UI changes that can't be
             * expressed declaratively. For example if you need to set a property or an attribute whose name is
             * not statically known, use {@link #setElementProperty} or {@link #setElementAttribute}
             * respectively.
             *
             * If you are implementing a custom renderer, you must implement this interface.
             *
             * The default Renderer implementation is `DomRenderer`. Also available is `WebWorkerRenderer`.
             */
            RootRenderer = (function () {
                function RootRenderer() {
                }
                return RootRenderer;
            }());
            exports_1("RootRenderer", RootRenderer);
        }
    }
});
