// angular2/ts/src/core/di/reflective_provider.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/collection', 'angular2/src/core/reflection/reflection', './reflective_key', './metadata', './reflective_exceptions', './forward_ref', './provider'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, collection_1, reflection_1, reflective_key_1, metadata_1, reflective_exceptions_1, forward_ref_1, provider_1;
    var ReflectiveDependency, _EMPTY_LIST, ResolvedReflectiveProvider_, ResolvedReflectiveFactory;
    /**
     * Resolve a single provider.
     */
    function resolveReflectiveFactory(provider) {
        var factoryFn;
        var resolvedDeps;
        if (lang_1.isPresent(provider.useClass)) {
            var useClass = forward_ref_1.resolveForwardRef(provider.useClass);
            factoryFn = reflection_1.reflector.factory(useClass);
            resolvedDeps = _dependenciesFor(useClass);
        }
        else if (lang_1.isPresent(provider.useExisting)) {
            factoryFn = function (aliasInstance) { return aliasInstance; };
            resolvedDeps = [ReflectiveDependency.fromKey(reflective_key_1.ReflectiveKey.get(provider.useExisting))];
        }
        else if (lang_1.isPresent(provider.useFactory)) {
            factoryFn = provider.useFactory;
            resolvedDeps = constructDependencies(provider.useFactory, provider.dependencies);
        }
        else {
            factoryFn = function () { return provider.useValue; };
            resolvedDeps = _EMPTY_LIST;
        }
        return new ResolvedReflectiveFactory(factoryFn, resolvedDeps);
    }
    exports_1("resolveReflectiveFactory", resolveReflectiveFactory);
    /**
     * Converts the {@link Provider} into {@link ResolvedProvider}.
     *
     * {@link Injector} internally only uses {@link ResolvedProvider}, {@link Provider} contains
     * convenience provider syntax.
     */
    function resolveReflectiveProvider(provider) {
        return new ResolvedReflectiveProvider_(reflective_key_1.ReflectiveKey.get(provider.token), [resolveReflectiveFactory(provider)], provider.multi);
    }
    exports_1("resolveReflectiveProvider", resolveReflectiveProvider);
    /**
     * Resolve a list of Providers.
     */
    function resolveReflectiveProviders(providers) {
        var normalized = _normalizeProviders(providers, []);
        var resolved = normalized.map(resolveReflectiveProvider);
        return collection_1.MapWrapper.values(mergeResolvedReflectiveProviders(resolved, new Map()));
    }
    exports_1("resolveReflectiveProviders", resolveReflectiveProviders);
    /**
     * Merges a list of ResolvedProviders into a list where
     * each key is contained exactly once and multi providers
     * have been merged.
     */
    function mergeResolvedReflectiveProviders(providers, normalizedProvidersMap) {
        for (var i = 0; i < providers.length; i++) {
            var provider = providers[i];
            var existing = normalizedProvidersMap.get(provider.key.id);
            if (lang_1.isPresent(existing)) {
                if (provider.multiProvider !== existing.multiProvider) {
                    throw new reflective_exceptions_1.MixingMultiProvidersWithRegularProvidersError(existing, provider);
                }
                if (provider.multiProvider) {
                    for (var j = 0; j < provider.resolvedFactories.length; j++) {
                        existing.resolvedFactories.push(provider.resolvedFactories[j]);
                    }
                }
                else {
                    normalizedProvidersMap.set(provider.key.id, provider);
                }
            }
            else {
                var resolvedProvider;
                if (provider.multiProvider) {
                    resolvedProvider = new ResolvedReflectiveProvider_(provider.key, collection_1.ListWrapper.clone(provider.resolvedFactories), provider.multiProvider);
                }
                else {
                    resolvedProvider = provider;
                }
                normalizedProvidersMap.set(provider.key.id, resolvedProvider);
            }
        }
        return normalizedProvidersMap;
    }
    exports_1("mergeResolvedReflectiveProviders", mergeResolvedReflectiveProviders);
    function _normalizeProviders(providers, res) {
        providers.forEach(function (b) {
            if (b instanceof lang_1.Type) {
                res.push(provider_1.provide(b, { useClass: b }));
            }
            else if (b instanceof provider_1.Provider) {
                res.push(b);
            }
            else if (b instanceof Array) {
                _normalizeProviders(b, res);
            }
            else if (b instanceof provider_1.ProviderBuilder) {
                throw new reflective_exceptions_1.InvalidProviderError(b.token);
            }
            else {
                throw new reflective_exceptions_1.InvalidProviderError(b);
            }
        });
        return res;
    }
    function constructDependencies(typeOrFunc, dependencies) {
        if (lang_1.isBlank(dependencies)) {
            return _dependenciesFor(typeOrFunc);
        }
        else {
            var params = dependencies.map(function (t) { return [t]; });
            return dependencies.map(function (t) { return _extractToken(typeOrFunc, t, params); });
        }
    }
    exports_1("constructDependencies", constructDependencies);
    function _dependenciesFor(typeOrFunc) {
        var params = reflection_1.reflector.parameters(typeOrFunc);
        if (lang_1.isBlank(params))
            return [];
        if (params.some(lang_1.isBlank)) {
            throw new reflective_exceptions_1.NoAnnotationError(typeOrFunc, params);
        }
        return params.map(function (p) { return _extractToken(typeOrFunc, p, params); });
    }
    function _extractToken(typeOrFunc, metadata /*any[] | any*/, params) {
        var depProps = [];
        var token = null;
        var optional = false;
        if (!lang_1.isArray(metadata)) {
            if (metadata instanceof metadata_1.InjectMetadata) {
                return _createDependency(metadata.token, optional, null, null, depProps);
            }
            else {
                return _createDependency(metadata, optional, null, null, depProps);
            }
        }
        var lowerBoundVisibility = null;
        var upperBoundVisibility = null;
        for (var i = 0; i < metadata.length; ++i) {
            var paramMetadata = metadata[i];
            if (paramMetadata instanceof lang_1.Type) {
                token = paramMetadata;
            }
            else if (paramMetadata instanceof metadata_1.InjectMetadata) {
                token = paramMetadata.token;
            }
            else if (paramMetadata instanceof metadata_1.OptionalMetadata) {
                optional = true;
            }
            else if (paramMetadata instanceof metadata_1.SelfMetadata) {
                upperBoundVisibility = paramMetadata;
            }
            else if (paramMetadata instanceof metadata_1.HostMetadata) {
                upperBoundVisibility = paramMetadata;
            }
            else if (paramMetadata instanceof metadata_1.SkipSelfMetadata) {
                lowerBoundVisibility = paramMetadata;
            }
            else if (paramMetadata instanceof metadata_1.DependencyMetadata) {
                if (lang_1.isPresent(paramMetadata.token)) {
                    token = paramMetadata.token;
                }
                depProps.push(paramMetadata);
            }
        }
        token = forward_ref_1.resolveForwardRef(token);
        if (lang_1.isPresent(token)) {
            return _createDependency(token, optional, lowerBoundVisibility, upperBoundVisibility, depProps);
        }
        else {
            throw new reflective_exceptions_1.NoAnnotationError(typeOrFunc, params);
        }
    }
    function _createDependency(token, optional, lowerBoundVisibility, upperBoundVisibility, depProps) {
        return new ReflectiveDependency(reflective_key_1.ReflectiveKey.get(token), optional, lowerBoundVisibility, upperBoundVisibility, depProps);
    }
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (reflection_1_1) {
                reflection_1 = reflection_1_1;
            },
            function (reflective_key_1_1) {
                reflective_key_1 = reflective_key_1_1;
            },
            function (metadata_1_1) {
                metadata_1 = metadata_1_1;
            },
            function (reflective_exceptions_1_1) {
                reflective_exceptions_1 = reflective_exceptions_1_1;
            },
            function (forward_ref_1_1) {
                forward_ref_1 = forward_ref_1_1;
            },
            function (provider_1_1) {
                provider_1 = provider_1_1;
            }],
        execute: function() {
            /**
             * `Dependency` is used by the framework to extend DI.
             * This is internal to Angular and should not be used directly.
             */
            ReflectiveDependency = (function () {
                function ReflectiveDependency(key, optional, lowerBoundVisibility, upperBoundVisibility, properties) {
                    this.key = key;
                    this.optional = optional;
                    this.lowerBoundVisibility = lowerBoundVisibility;
                    this.upperBoundVisibility = upperBoundVisibility;
                    this.properties = properties;
                }
                ReflectiveDependency.fromKey = function (key) {
                    return new ReflectiveDependency(key, false, null, null, []);
                };
                return ReflectiveDependency;
            }());
            exports_1("ReflectiveDependency", ReflectiveDependency);
            _EMPTY_LIST = lang_1.CONST_EXPR([]);
            ResolvedReflectiveProvider_ = (function () {
                function ResolvedReflectiveProvider_(key, resolvedFactories, multiProvider) {
                    this.key = key;
                    this.resolvedFactories = resolvedFactories;
                    this.multiProvider = multiProvider;
                }
                Object.defineProperty(ResolvedReflectiveProvider_.prototype, "resolvedFactory", {
                    get: function () { return this.resolvedFactories[0]; },
                    enumerable: true,
                    configurable: true
                });
                return ResolvedReflectiveProvider_;
            }());
            exports_1("ResolvedReflectiveProvider_", ResolvedReflectiveProvider_);
            /**
             * An internal resolved representation of a factory function created by resolving {@link Provider}.
             */
            ResolvedReflectiveFactory = (function () {
                function ResolvedReflectiveFactory(
                    /**
                     * Factory function which can return an instance of an object represented by a key.
                     */
                    factory, 
                    /**
                     * Arguments (dependencies) to the `factory` function.
                     */
                    dependencies) {
                    this.factory = factory;
                    this.dependencies = dependencies;
                }
                return ResolvedReflectiveFactory;
            }());
            exports_1("ResolvedReflectiveFactory", ResolvedReflectiveFactory);
        }
    }
});
