// angular2/ts/src/core/change_detection/change_detection_util.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/collection'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, collection_1;
    var uninitialized, WrappedValue, ValueUnwrapper, SimpleChange;
    function devModeEqual(a, b) {
        if (collection_1.isListLikeIterable(a) && collection_1.isListLikeIterable(b)) {
            return collection_1.areIterablesEqual(a, b, devModeEqual);
        }
        else if (!collection_1.isListLikeIterable(a) && !lang_1.isPrimitive(a) && !collection_1.isListLikeIterable(b) &&
            !lang_1.isPrimitive(b)) {
            return true;
        }
        else {
            return lang_1.looseIdentical(a, b);
        }
    }
    exports_1("devModeEqual", devModeEqual);
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
                exports_1({
                    "looseIdentical": lang_1_1["looseIdentical"]
                });
            },
            function (collection_1_1) {
                collection_1 = collection_1_1;
            }],
        execute: function() {
            exports_1("uninitialized", uninitialized = lang_1.CONST_EXPR(new Object()));
            /**
             * Indicates that the result of a {@link PipeMetadata} transformation has changed even though the
             * reference
             * has not changed.
             *
             * The wrapped value will be unwrapped by change detection, and the unwrapped value will be stored.
             *
             * Example:
             *
             * ```
             * if (this._latestValue === this._latestReturnedValue) {
             *    return this._latestReturnedValue;
             *  } else {
             *    this._latestReturnedValue = this._latestValue;
             *    return WrappedValue.wrap(this._latestValue); // this will force update
             *  }
             * ```
             */
            WrappedValue = (function () {
                function WrappedValue(wrapped) {
                    this.wrapped = wrapped;
                }
                WrappedValue.wrap = function (value) { return new WrappedValue(value); };
                return WrappedValue;
            }());
            exports_1("WrappedValue", WrappedValue);
            /**
             * Helper class for unwrapping WrappedValue s
             */
            ValueUnwrapper = (function () {
                function ValueUnwrapper() {
                    this.hasWrappedValue = false;
                }
                ValueUnwrapper.prototype.unwrap = function (value) {
                    if (value instanceof WrappedValue) {
                        this.hasWrappedValue = true;
                        return value.wrapped;
                    }
                    return value;
                };
                ValueUnwrapper.prototype.reset = function () { this.hasWrappedValue = false; };
                return ValueUnwrapper;
            }());
            exports_1("ValueUnwrapper", ValueUnwrapper);
            /**
             * Represents a basic change from a previous to a new value.
             */
            SimpleChange = (function () {
                function SimpleChange(previousValue, currentValue) {
                    this.previousValue = previousValue;
                    this.currentValue = currentValue;
                }
                /**
                 * Check whether the new value is the first value assigned.
                 */
                SimpleChange.prototype.isFirstChange = function () { return this.previousValue === uninitialized; };
                return SimpleChange;
            }());
            exports_1("SimpleChange", SimpleChange);
        }
    }
});
