// angular2/ts/src/compiler/view_compiler/compile_query.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/collection', '../output/output_ast', '../identifiers', './util'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, collection_1, o, identifiers_1, util_1;
    var ViewQueryValues, CompileQuery;
    function createQueryValues(viewValues) {
        return collection_1.ListWrapper.flatten(viewValues.values.map(function (entry) {
            if (entry instanceof ViewQueryValues) {
                return mapNestedViews(entry.view.declarationElement.appElement, entry.view, createQueryValues(entry));
            }
            else {
                return entry;
            }
        }));
    }
    function mapNestedViews(declarationAppElement, view, expressions) {
        var adjustedExpressions = expressions.map(function (expr) {
            return o.replaceVarInExpression(o.THIS_EXPR.name, o.variable('nestedView'), expr);
        });
        return declarationAppElement.callMethod('mapNestedViews', [
            o.variable(view.className),
            o.fn([new o.FnParam('nestedView', view.classType)], [new o.ReturnStatement(o.literalArr(adjustedExpressions))])
        ]);
    }
    function createQueryList(query, directiveInstance, propertyName, compileView) {
        compileView.fields.push(new o.ClassField(propertyName, o.importType(identifiers_1.Identifiers.QueryList), [o.StmtModifier.Private]));
        var expr = o.THIS_EXPR.prop(propertyName);
        compileView.createMethod.addStmt(o.THIS_EXPR.prop(propertyName)
            .set(o.importExpr(identifiers_1.Identifiers.QueryList).instantiate([]))
            .toStmt());
        return expr;
    }
    exports_1("createQueryList", createQueryList);
    function addQueryToTokenMap(map, query) {
        query.meta.selectors.forEach(function (selector) {
            var entry = map.get(selector);
            if (lang_1.isBlank(entry)) {
                entry = [];
                map.add(selector, entry);
            }
            entry.push(query);
        });
    }
    exports_1("addQueryToTokenMap", addQueryToTokenMap);
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (o_1) {
                o = o_1;
            },
            function (identifiers_1_1) {
                identifiers_1 = identifiers_1_1;
            },
            function (util_1_1) {
                util_1 = util_1_1;
            }],
        execute: function() {
            ViewQueryValues = (function () {
                function ViewQueryValues(view, values) {
                    this.view = view;
                    this.values = values;
                }
                return ViewQueryValues;
            }());
            CompileQuery = (function () {
                function CompileQuery(meta, queryList, ownerDirectiveExpression, view) {
                    this.meta = meta;
                    this.queryList = queryList;
                    this.ownerDirectiveExpression = ownerDirectiveExpression;
                    this.view = view;
                    this._values = new ViewQueryValues(view, []);
                }
                CompileQuery.prototype.addValue = function (value, view) {
                    var currentView = view;
                    var elPath = [];
                    while (lang_1.isPresent(currentView) && currentView !== this.view) {
                        var parentEl = currentView.declarationElement;
                        elPath.unshift(parentEl);
                        currentView = parentEl.view;
                    }
                    var queryListForDirtyExpr = util_1.getPropertyInView(this.queryList, view, this.view);
                    var viewValues = this._values;
                    elPath.forEach(function (el) {
                        var last = viewValues.values.length > 0 ? viewValues.values[viewValues.values.length - 1] : null;
                        if (last instanceof ViewQueryValues && last.view === el.embeddedView) {
                            viewValues = last;
                        }
                        else {
                            var newViewValues = new ViewQueryValues(el.embeddedView, []);
                            viewValues.values.push(newViewValues);
                            viewValues = newViewValues;
                        }
                    });
                    viewValues.values.push(value);
                    if (elPath.length > 0) {
                        view.dirtyParentQueriesMethod.addStmt(queryListForDirtyExpr.callMethod('setDirty', []).toStmt());
                    }
                };
                CompileQuery.prototype.afterChildren = function (targetMethod) {
                    var values = createQueryValues(this._values);
                    var updateStmts = [this.queryList.callMethod('reset', [o.literalArr(values)]).toStmt()];
                    if (lang_1.isPresent(this.ownerDirectiveExpression)) {
                        var valueExpr = this.meta.first ? this.queryList.prop('first') : this.queryList;
                        updateStmts.push(this.ownerDirectiveExpression.prop(this.meta.propertyName).set(valueExpr).toStmt());
                    }
                    if (!this.meta.first) {
                        updateStmts.push(this.queryList.callMethod('notifyOnChanges', []).toStmt());
                    }
                    targetMethod.addStmt(new o.IfStmt(this.queryList.prop('dirty'), updateStmts));
                };
                return CompileQuery;
            }());
            exports_1("CompileQuery", CompileQuery);
        }
    }
});
