// angular2/ts/src/compiler/view_compiler/constants.js
System.register(['angular2/src/facade/lang', '../compile_metadata', 'angular2/src/core/change_detection/change_detection', 'angular2/src/core/metadata/view', 'angular2/src/core/linker/view_type', '../output/output_ast', '../identifiers'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, compile_metadata_1, change_detection_1, view_1, view_type_1, o, identifiers_1;
    var ViewTypeEnum, ViewEncapsulationEnum, ChangeDetectorStateEnum, ChangeDetectionStrategyEnum, ViewConstructorVars, ViewProperties, EventHandlerVars, InjectMethodVars, DetectChangesVars;
    function _enumExpression(classIdentifier, value) {
        if (lang_1.isBlank(value))
            return o.NULL_EXPR;
        var name = lang_1.resolveEnumToken(classIdentifier.runtime, value);
        return o.importExpr(new compile_metadata_1.CompileIdentifierMetadata({
            name: classIdentifier.name + "." + name,
            moduleUrl: classIdentifier.moduleUrl,
            runtime: value
        }));
    }
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (compile_metadata_1_1) {
                compile_metadata_1 = compile_metadata_1_1;
            },
            function (change_detection_1_1) {
                change_detection_1 = change_detection_1_1;
            },
            function (view_1_1) {
                view_1 = view_1_1;
            },
            function (view_type_1_1) {
                view_type_1 = view_type_1_1;
            },
            function (o_1) {
                o = o_1;
            },
            function (identifiers_1_1) {
                identifiers_1 = identifiers_1_1;
            }],
        execute: function() {
            ViewTypeEnum = (function () {
                function ViewTypeEnum() {
                }
                ViewTypeEnum.fromValue = function (value) {
                    return _enumExpression(identifiers_1.Identifiers.ViewType, value);
                };
                ViewTypeEnum.HOST = ViewTypeEnum.fromValue(view_type_1.ViewType.HOST);
                ViewTypeEnum.COMPONENT = ViewTypeEnum.fromValue(view_type_1.ViewType.COMPONENT);
                ViewTypeEnum.EMBEDDED = ViewTypeEnum.fromValue(view_type_1.ViewType.EMBEDDED);
                return ViewTypeEnum;
            }());
            exports_1("ViewTypeEnum", ViewTypeEnum);
            ViewEncapsulationEnum = (function () {
                function ViewEncapsulationEnum() {
                }
                ViewEncapsulationEnum.fromValue = function (value) {
                    return _enumExpression(identifiers_1.Identifiers.ViewEncapsulation, value);
                };
                ViewEncapsulationEnum.Emulated = ViewEncapsulationEnum.fromValue(view_1.ViewEncapsulation.Emulated);
                ViewEncapsulationEnum.Native = ViewEncapsulationEnum.fromValue(view_1.ViewEncapsulation.Native);
                ViewEncapsulationEnum.None = ViewEncapsulationEnum.fromValue(view_1.ViewEncapsulation.None);
                return ViewEncapsulationEnum;
            }());
            exports_1("ViewEncapsulationEnum", ViewEncapsulationEnum);
            ChangeDetectorStateEnum = (function () {
                function ChangeDetectorStateEnum() {
                }
                ChangeDetectorStateEnum.fromValue = function (value) {
                    return _enumExpression(identifiers_1.Identifiers.ChangeDetectorState, value);
                };
                ChangeDetectorStateEnum.NeverChecked = ChangeDetectorStateEnum.fromValue(change_detection_1.ChangeDetectorState.NeverChecked);
                ChangeDetectorStateEnum.CheckedBefore = ChangeDetectorStateEnum.fromValue(change_detection_1.ChangeDetectorState.CheckedBefore);
                ChangeDetectorStateEnum.Errored = ChangeDetectorStateEnum.fromValue(change_detection_1.ChangeDetectorState.Errored);
                return ChangeDetectorStateEnum;
            }());
            exports_1("ChangeDetectorStateEnum", ChangeDetectorStateEnum);
            ChangeDetectionStrategyEnum = (function () {
                function ChangeDetectionStrategyEnum() {
                }
                ChangeDetectionStrategyEnum.fromValue = function (value) {
                    return _enumExpression(identifiers_1.Identifiers.ChangeDetectionStrategy, value);
                };
                ChangeDetectionStrategyEnum.CheckOnce = ChangeDetectionStrategyEnum.fromValue(change_detection_1.ChangeDetectionStrategy.CheckOnce);
                ChangeDetectionStrategyEnum.Checked = ChangeDetectionStrategyEnum.fromValue(change_detection_1.ChangeDetectionStrategy.Checked);
                ChangeDetectionStrategyEnum.CheckAlways = ChangeDetectionStrategyEnum.fromValue(change_detection_1.ChangeDetectionStrategy.CheckAlways);
                ChangeDetectionStrategyEnum.Detached = ChangeDetectionStrategyEnum.fromValue(change_detection_1.ChangeDetectionStrategy.Detached);
                ChangeDetectionStrategyEnum.OnPush = ChangeDetectionStrategyEnum.fromValue(change_detection_1.ChangeDetectionStrategy.OnPush);
                ChangeDetectionStrategyEnum.Default = ChangeDetectionStrategyEnum.fromValue(change_detection_1.ChangeDetectionStrategy.Default);
                return ChangeDetectionStrategyEnum;
            }());
            exports_1("ChangeDetectionStrategyEnum", ChangeDetectionStrategyEnum);
            ViewConstructorVars = (function () {
                function ViewConstructorVars() {
                }
                ViewConstructorVars.viewUtils = o.variable('viewUtils');
                ViewConstructorVars.parentInjector = o.variable('parentInjector');
                ViewConstructorVars.declarationEl = o.variable('declarationEl');
                return ViewConstructorVars;
            }());
            exports_1("ViewConstructorVars", ViewConstructorVars);
            ViewProperties = (function () {
                function ViewProperties() {
                }
                ViewProperties.renderer = o.THIS_EXPR.prop('renderer');
                ViewProperties.projectableNodes = o.THIS_EXPR.prop('projectableNodes');
                ViewProperties.viewUtils = o.THIS_EXPR.prop('viewUtils');
                return ViewProperties;
            }());
            exports_1("ViewProperties", ViewProperties);
            EventHandlerVars = (function () {
                function EventHandlerVars() {
                }
                EventHandlerVars.event = o.variable('$event');
                return EventHandlerVars;
            }());
            exports_1("EventHandlerVars", EventHandlerVars);
            InjectMethodVars = (function () {
                function InjectMethodVars() {
                }
                InjectMethodVars.token = o.variable('token');
                InjectMethodVars.requestNodeIndex = o.variable('requestNodeIndex');
                InjectMethodVars.notFoundResult = o.variable('notFoundResult');
                return InjectMethodVars;
            }());
            exports_1("InjectMethodVars", InjectMethodVars);
            DetectChangesVars = (function () {
                function DetectChangesVars() {
                }
                DetectChangesVars.throwOnChange = o.variable("throwOnChange");
                DetectChangesVars.changes = o.variable("changes");
                DetectChangesVars.changed = o.variable("changed");
                DetectChangesVars.valUnwrapper = o.variable("valUnwrapper");
                return DetectChangesVars;
            }());
            exports_1("DetectChangesVars", DetectChangesVars);
        }
    }
});
