// angular2/ts/src/compiler/view_compiler/property_binder.js
System.register(['../output/output_ast', '../identifiers', './constants', '../template_ast', 'angular2/src/facade/lang', 'angular2/src/core/metadata/lifecycle_hooks', 'angular2/src/core/change_detection/constants', '../util', './expression_converter', './compile_binding'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var o, identifiers_1, constants_1, template_ast_1, lang_1, lifecycle_hooks_1, constants_2, util_1, expression_converter_1, compile_binding_1;
    function createBindFieldExpr(exprIndex) {
        return o.THIS_EXPR.prop("_expr_" + exprIndex);
    }
    function createCurrValueExpr(exprIndex) {
        return o.variable("currVal_" + exprIndex);
    }
    function bind(view, currValExpr, fieldExpr, parsedExpression, context, actions, method) {
        var checkExpression = expression_converter_1.convertCdExpressionToIr(view, context, parsedExpression, constants_1.DetectChangesVars.valUnwrapper);
        if (lang_1.isBlank(checkExpression.expression)) {
            // e.g. an empty expression was given
            return;
        }
        view.fields.push(new o.ClassField(fieldExpr.name, null, [o.StmtModifier.Private]));
        view.createMethod.addStmt(o.THIS_EXPR.prop(fieldExpr.name).set(o.importExpr(identifiers_1.Identifiers.uninitialized)).toStmt());
        if (checkExpression.needsValueUnwrapper) {
            var initValueUnwrapperStmt = constants_1.DetectChangesVars.valUnwrapper.callMethod('reset', []).toStmt();
            method.addStmt(initValueUnwrapperStmt);
        }
        method.addStmt(currValExpr.set(checkExpression.expression).toDeclStmt(null, [o.StmtModifier.Final]));
        var condition = o.importExpr(identifiers_1.Identifiers.checkBinding)
            .callFn([constants_1.DetectChangesVars.throwOnChange, fieldExpr, currValExpr]);
        if (checkExpression.needsValueUnwrapper) {
            condition = constants_1.DetectChangesVars.valUnwrapper.prop('hasWrappedValue').or(condition);
        }
        method.addStmt(new o.IfStmt(condition, actions.concat([o.THIS_EXPR.prop(fieldExpr.name).set(currValExpr).toStmt()])));
    }
    function bindRenderText(boundText, compileNode, view) {
        var bindingIndex = view.bindings.length;
        view.bindings.push(new compile_binding_1.CompileBinding(compileNode, boundText));
        var currValExpr = createCurrValueExpr(bindingIndex);
        var valueField = createBindFieldExpr(bindingIndex);
        view.detectChangesRenderPropertiesMethod.resetDebugInfo(compileNode.nodeIndex, boundText);
        bind(view, currValExpr, valueField, boundText.value, o.THIS_EXPR.prop('context'), [
            o.THIS_EXPR.prop('renderer')
                .callMethod('setText', [compileNode.renderNode, currValExpr])
                .toStmt()
        ], view.detectChangesRenderPropertiesMethod);
    }
    exports_1("bindRenderText", bindRenderText);
    function bindAndWriteToRenderer(boundProps, context, compileElement) {
        var view = compileElement.view;
        var renderNode = compileElement.renderNode;
        boundProps.forEach(function (boundProp) {
            var bindingIndex = view.bindings.length;
            view.bindings.push(new compile_binding_1.CompileBinding(compileElement, boundProp));
            view.detectChangesRenderPropertiesMethod.resetDebugInfo(compileElement.nodeIndex, boundProp);
            var fieldExpr = createBindFieldExpr(bindingIndex);
            var currValExpr = createCurrValueExpr(bindingIndex);
            var renderMethod;
            var renderValue = currValExpr;
            var updateStmts = [];
            switch (boundProp.type) {
                case template_ast_1.PropertyBindingType.Property:
                    renderMethod = 'setElementProperty';
                    if (view.genConfig.logBindingUpdate) {
                        updateStmts.push(logBindingUpdateStmt(renderNode, boundProp.name, currValExpr));
                    }
                    break;
                case template_ast_1.PropertyBindingType.Attribute:
                    renderMethod = 'setElementAttribute';
                    renderValue =
                        renderValue.isBlank().conditional(o.NULL_EXPR, renderValue.callMethod('toString', []));
                    break;
                case template_ast_1.PropertyBindingType.Class:
                    renderMethod = 'setElementClass';
                    break;
                case template_ast_1.PropertyBindingType.Style:
                    renderMethod = 'setElementStyle';
                    var strValue = renderValue.callMethod('toString', []);
                    if (lang_1.isPresent(boundProp.unit)) {
                        strValue = strValue.plus(o.literal(boundProp.unit));
                    }
                    renderValue = renderValue.isBlank().conditional(o.NULL_EXPR, strValue);
                    break;
            }
            updateStmts.push(o.THIS_EXPR.prop('renderer')
                .callMethod(renderMethod, [renderNode, o.literal(boundProp.name), renderValue])
                .toStmt());
            bind(view, currValExpr, fieldExpr, boundProp.value, context, updateStmts, view.detectChangesRenderPropertiesMethod);
        });
    }
    function bindRenderInputs(boundProps, compileElement) {
        bindAndWriteToRenderer(boundProps, o.THIS_EXPR.prop('context'), compileElement);
    }
    exports_1("bindRenderInputs", bindRenderInputs);
    function bindDirectiveHostProps(directiveAst, directiveInstance, compileElement) {
        bindAndWriteToRenderer(directiveAst.hostProperties, directiveInstance, compileElement);
    }
    exports_1("bindDirectiveHostProps", bindDirectiveHostProps);
    function bindDirectiveInputs(directiveAst, directiveInstance, compileElement) {
        if (directiveAst.inputs.length === 0) {
            return;
        }
        var view = compileElement.view;
        var detectChangesInInputsMethod = view.detectChangesInInputsMethod;
        detectChangesInInputsMethod.resetDebugInfo(compileElement.nodeIndex, compileElement.sourceAst);
        var lifecycleHooks = directiveAst.directive.lifecycleHooks;
        var calcChangesMap = lifecycleHooks.indexOf(lifecycle_hooks_1.LifecycleHooks.OnChanges) !== -1;
        var isOnPushComp = directiveAst.directive.isComponent &&
            !constants_2.isDefaultChangeDetectionStrategy(directiveAst.directive.changeDetection);
        if (calcChangesMap) {
            detectChangesInInputsMethod.addStmt(constants_1.DetectChangesVars.changes.set(o.NULL_EXPR).toStmt());
        }
        if (isOnPushComp) {
            detectChangesInInputsMethod.addStmt(constants_1.DetectChangesVars.changed.set(o.literal(false)).toStmt());
        }
        directiveAst.inputs.forEach(function (input) {
            var bindingIndex = view.bindings.length;
            view.bindings.push(new compile_binding_1.CompileBinding(compileElement, input));
            detectChangesInInputsMethod.resetDebugInfo(compileElement.nodeIndex, input);
            var fieldExpr = createBindFieldExpr(bindingIndex);
            var currValExpr = createCurrValueExpr(bindingIndex);
            var statements = [directiveInstance.prop(input.directiveName).set(currValExpr).toStmt()];
            if (calcChangesMap) {
                statements.push(new o.IfStmt(constants_1.DetectChangesVars.changes.identical(o.NULL_EXPR), [
                    constants_1.DetectChangesVars.changes.set(o.literalMap([], new o.MapType(o.importType(identifiers_1.Identifiers.SimpleChange))))
                        .toStmt()
                ]));
                statements.push(constants_1.DetectChangesVars.changes.key(o.literal(input.directiveName))
                    .set(o.importExpr(identifiers_1.Identifiers.SimpleChange).instantiate([fieldExpr, currValExpr]))
                    .toStmt());
            }
            if (isOnPushComp) {
                statements.push(constants_1.DetectChangesVars.changed.set(o.literal(true)).toStmt());
            }
            if (view.genConfig.logBindingUpdate) {
                statements.push(logBindingUpdateStmt(compileElement.renderNode, input.directiveName, currValExpr));
            }
            bind(view, currValExpr, fieldExpr, input.value, o.THIS_EXPR.prop('context'), statements, detectChangesInInputsMethod);
        });
        if (isOnPushComp) {
            detectChangesInInputsMethod.addStmt(new o.IfStmt(constants_1.DetectChangesVars.changed, [
                compileElement.appElement.prop('componentView')
                    .callMethod('markAsCheckOnce', [])
                    .toStmt()
            ]));
        }
    }
    exports_1("bindDirectiveInputs", bindDirectiveInputs);
    function logBindingUpdateStmt(renderNode, propName, value) {
        return o.THIS_EXPR.prop('renderer')
            .callMethod('setBindingDebugInfo', [
            renderNode,
            o.literal("ng-reflect-" + util_1.camelCaseToDashCase(propName)),
            value.isBlank().conditional(o.NULL_EXPR, value.callMethod('toString', []))
        ])
            .toStmt();
    }
    return {
        setters:[
            function (o_1) {
                o = o_1;
            },
            function (identifiers_1_1) {
                identifiers_1 = identifiers_1_1;
            },
            function (constants_1_1) {
                constants_1 = constants_1_1;
            },
            function (template_ast_1_1) {
                template_ast_1 = template_ast_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (lifecycle_hooks_1_1) {
                lifecycle_hooks_1 = lifecycle_hooks_1_1;
            },
            function (constants_2_1) {
                constants_2 = constants_2_1;
            },
            function (util_1_1) {
                util_1 = util_1_1;
            },
            function (expression_converter_1_1) {
                expression_converter_1 = expression_converter_1_1;
            },
            function (compile_binding_1_1) {
                compile_binding_1 = compile_binding_1_1;
            }],
        execute: function() {
        }
    }
});
