// angular2/ts/src/compiler/config.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/exceptions', './identifiers'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, exceptions_1, identifiers_1;
    var CompilerConfig, RenderTypes, DefaultRenderTypes;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (identifiers_1_1) {
                identifiers_1 = identifiers_1_1;
            }],
        execute: function() {
            CompilerConfig = (function () {
                function CompilerConfig(genDebugInfo, logBindingUpdate, useJit, renderTypes) {
                    if (renderTypes === void 0) { renderTypes = null; }
                    this.genDebugInfo = genDebugInfo;
                    this.logBindingUpdate = logBindingUpdate;
                    this.useJit = useJit;
                    if (lang_1.isBlank(renderTypes)) {
                        renderTypes = new DefaultRenderTypes();
                    }
                    this.renderTypes = renderTypes;
                }
                return CompilerConfig;
            }());
            exports_1("CompilerConfig", CompilerConfig);
            /**
             * Types used for the renderer.
             * Can be replaced to specialize the generated output to a specific renderer
             * to help tree shaking.
             */
            RenderTypes = (function () {
                function RenderTypes() {
                }
                Object.defineProperty(RenderTypes.prototype, "renderer", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderTypes.prototype, "renderText", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderTypes.prototype, "renderElement", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderTypes.prototype, "renderComment", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderTypes.prototype, "renderNode", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(RenderTypes.prototype, "renderEvent", {
                    get: function () { return exceptions_1.unimplemented(); },
                    enumerable: true,
                    configurable: true
                });
                return RenderTypes;
            }());
            exports_1("RenderTypes", RenderTypes);
            DefaultRenderTypes = (function () {
                function DefaultRenderTypes() {
                    this.renderer = identifiers_1.Identifiers.Renderer;
                    this.renderText = null;
                    this.renderElement = null;
                    this.renderComment = null;
                    this.renderNode = null;
                    this.renderEvent = null;
                }
                return DefaultRenderTypes;
            }());
            exports_1("DefaultRenderTypes", DefaultRenderTypes);
        }
    }
});
