// angular2/ts/src/compiler/identifiers.js
System.register(['./compile_metadata', 'angular2/src/core/linker/view', 'angular2/src/core/linker/debug_context', 'angular2/src/core/linker/view_utils', 'angular2/src/core/change_detection/change_detection', 'angular2/src/core/linker/element', 'angular2/src/core/linker/element_ref', 'angular2/src/core/linker/view_container_ref', 'angular2/src/core/render/api', 'angular2/src/core/metadata/view', 'angular2/src/core/linker/view_type', 'angular2/src/core/linker', 'angular2/src/core/di/injector', 'angular2/src/core/linker/template_ref', './util'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var compile_metadata_1, view_1, debug_context_1, view_utils_1, change_detection_1, element_1, element_ref_1, view_container_ref_1, api_1, view_2, view_type_1, linker_1, injector_1, template_ref_1, util_1;
    var APP_VIEW_MODULE_URL, VIEW_UTILS_MODULE_URL, CD_MODULE_URL, impViewUtils, impAppView, impDebugContext, impAppElement, impElementRef, impViewContainerRef, impChangeDetectorRef, impRenderComponentType, impQueryList, impTemplateRef, impTemplateRef_, impValueUnwrapper, impInjector, impViewEncapsulation, impViewType, impChangeDetectionStrategy, impStaticNodeDebugInfo, impRenderer, impSimpleChange, impUninitialized, impChangeDetectorState, impFlattenNestedViewRenderNodes, impDevModeEqual, impInterpolate, impCheckBinding, impCastByValue, Identifiers;
    function identifierToken(identifier) {
        return new compile_metadata_1.CompileTokenMetadata({ identifier: identifier });
    }
    exports_1("identifierToken", identifierToken);
    return {
        setters:[
            function (compile_metadata_1_1) {
                compile_metadata_1 = compile_metadata_1_1;
            },
            function (view_1_1) {
                view_1 = view_1_1;
            },
            function (debug_context_1_1) {
                debug_context_1 = debug_context_1_1;
            },
            function (view_utils_1_1) {
                view_utils_1 = view_utils_1_1;
            },
            function (change_detection_1_1) {
                change_detection_1 = change_detection_1_1;
            },
            function (element_1_1) {
                element_1 = element_1_1;
            },
            function (element_ref_1_1) {
                element_ref_1 = element_ref_1_1;
            },
            function (view_container_ref_1_1) {
                view_container_ref_1 = view_container_ref_1_1;
            },
            function (api_1_1) {
                api_1 = api_1_1;
            },
            function (view_2_1) {
                view_2 = view_2_1;
            },
            function (view_type_1_1) {
                view_type_1 = view_type_1_1;
            },
            function (linker_1_1) {
                linker_1 = linker_1_1;
            },
            function (injector_1_1) {
                injector_1 = injector_1_1;
            },
            function (template_ref_1_1) {
                template_ref_1 = template_ref_1_1;
            },
            function (util_1_1) {
                util_1 = util_1_1;
            }],
        execute: function() {
            APP_VIEW_MODULE_URL = 'asset:angular2/lib/src/core/linker/view' + util_1.MODULE_SUFFIX;
            VIEW_UTILS_MODULE_URL = 'asset:angular2/lib/src/core/linker/view_utils' + util_1.MODULE_SUFFIX;
            CD_MODULE_URL = 'asset:angular2/lib/src/core/change_detection/change_detection' + util_1.MODULE_SUFFIX;
            // Reassign the imports to different variables so we can
            // define static variables with the name of the import.
            // (only needed for Dart).
            impViewUtils = view_utils_1.ViewUtils;
            impAppView = view_1.AppView;
            impDebugContext = debug_context_1.DebugContext;
            impAppElement = element_1.AppElement;
            impElementRef = element_ref_1.ElementRef;
            impViewContainerRef = view_container_ref_1.ViewContainerRef;
            impChangeDetectorRef = change_detection_1.ChangeDetectorRef;
            impRenderComponentType = api_1.RenderComponentType;
            impQueryList = linker_1.QueryList;
            impTemplateRef = template_ref_1.TemplateRef;
            impTemplateRef_ = template_ref_1.TemplateRef_;
            impValueUnwrapper = change_detection_1.ValueUnwrapper;
            impInjector = injector_1.Injector;
            impViewEncapsulation = view_2.ViewEncapsulation;
            impViewType = view_type_1.ViewType;
            impChangeDetectionStrategy = change_detection_1.ChangeDetectionStrategy;
            impStaticNodeDebugInfo = debug_context_1.StaticNodeDebugInfo;
            impRenderer = api_1.Renderer;
            impSimpleChange = change_detection_1.SimpleChange;
            impUninitialized = change_detection_1.uninitialized;
            impChangeDetectorState = change_detection_1.ChangeDetectorState;
            impFlattenNestedViewRenderNodes = view_utils_1.flattenNestedViewRenderNodes;
            impDevModeEqual = change_detection_1.devModeEqual;
            impInterpolate = view_utils_1.interpolate;
            impCheckBinding = view_utils_1.checkBinding;
            impCastByValue = view_utils_1.castByValue;
            Identifiers = (function () {
                function Identifiers() {
                }
                Identifiers.ViewUtils = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'ViewUtils',
                    moduleUrl: 'asset:angular2/lib/src/core/linker/view_utils' + util_1.MODULE_SUFFIX,
                    runtime: impViewUtils
                });
                Identifiers.AppView = new compile_metadata_1.CompileIdentifierMetadata({ name: 'AppView', moduleUrl: APP_VIEW_MODULE_URL, runtime: impAppView });
                Identifiers.AppElement = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'AppElement',
                    moduleUrl: 'asset:angular2/lib/src/core/linker/element' + util_1.MODULE_SUFFIX,
                    runtime: impAppElement
                });
                Identifiers.ElementRef = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'ElementRef',
                    moduleUrl: 'asset:angular2/lib/src/core/linker/element_ref' + util_1.MODULE_SUFFIX,
                    runtime: impElementRef
                });
                Identifiers.ViewContainerRef = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'ViewContainerRef',
                    moduleUrl: 'asset:angular2/lib/src/core/linker/view_container_ref' + util_1.MODULE_SUFFIX,
                    runtime: impViewContainerRef
                });
                Identifiers.ChangeDetectorRef = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'ChangeDetectorRef',
                    moduleUrl: 'asset:angular2/lib/src/core/change_detection/change_detector_ref' + util_1.MODULE_SUFFIX,
                    runtime: impChangeDetectorRef
                });
                Identifiers.RenderComponentType = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'RenderComponentType',
                    moduleUrl: 'asset:angular2/lib/src/core/render/api' + util_1.MODULE_SUFFIX,
                    runtime: impRenderComponentType
                });
                Identifiers.QueryList = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'QueryList',
                    moduleUrl: 'asset:angular2/lib/src/core/linker/query_list' + util_1.MODULE_SUFFIX,
                    runtime: impQueryList
                });
                Identifiers.TemplateRef = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'TemplateRef',
                    moduleUrl: 'asset:angular2/lib/src/core/linker/template_ref' + util_1.MODULE_SUFFIX,
                    runtime: impTemplateRef
                });
                Identifiers.TemplateRef_ = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'TemplateRef_',
                    moduleUrl: 'asset:angular2/lib/src/core/linker/template_ref' + util_1.MODULE_SUFFIX,
                    runtime: impTemplateRef_
                });
                Identifiers.ValueUnwrapper = new compile_metadata_1.CompileIdentifierMetadata({ name: 'ValueUnwrapper', moduleUrl: CD_MODULE_URL, runtime: impValueUnwrapper });
                Identifiers.Injector = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'Injector',
                    moduleUrl: "asset:angular2/lib/src/core/di/injector" + util_1.MODULE_SUFFIX,
                    runtime: impInjector
                });
                Identifiers.ViewEncapsulation = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'ViewEncapsulation',
                    moduleUrl: 'asset:angular2/lib/src/core/metadata/view' + util_1.MODULE_SUFFIX,
                    runtime: impViewEncapsulation
                });
                Identifiers.ViewType = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'ViewType',
                    moduleUrl: "asset:angular2/lib/src/core/linker/view_type" + util_1.MODULE_SUFFIX,
                    runtime: impViewType
                });
                Identifiers.ChangeDetectionStrategy = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'ChangeDetectionStrategy',
                    moduleUrl: CD_MODULE_URL,
                    runtime: impChangeDetectionStrategy
                });
                Identifiers.StaticNodeDebugInfo = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'StaticNodeDebugInfo',
                    moduleUrl: "asset:angular2/lib/src/core/linker/debug_context" + util_1.MODULE_SUFFIX,
                    runtime: impStaticNodeDebugInfo
                });
                Identifiers.DebugContext = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'DebugContext',
                    moduleUrl: "asset:angular2/lib/src/core/linker/debug_context" + util_1.MODULE_SUFFIX,
                    runtime: impDebugContext
                });
                Identifiers.Renderer = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'Renderer',
                    moduleUrl: 'asset:angular2/lib/src/core/render/api' + util_1.MODULE_SUFFIX,
                    runtime: impRenderer
                });
                Identifiers.SimpleChange = new compile_metadata_1.CompileIdentifierMetadata({ name: 'SimpleChange', moduleUrl: CD_MODULE_URL, runtime: impSimpleChange });
                Identifiers.uninitialized = new compile_metadata_1.CompileIdentifierMetadata({ name: 'uninitialized', moduleUrl: CD_MODULE_URL, runtime: impUninitialized });
                Identifiers.ChangeDetectorState = new compile_metadata_1.CompileIdentifierMetadata({ name: 'ChangeDetectorState', moduleUrl: CD_MODULE_URL, runtime: impChangeDetectorState });
                Identifiers.checkBinding = new compile_metadata_1.CompileIdentifierMetadata({ name: 'checkBinding', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: impCheckBinding });
                Identifiers.flattenNestedViewRenderNodes = new compile_metadata_1.CompileIdentifierMetadata({
                    name: 'flattenNestedViewRenderNodes',
                    moduleUrl: VIEW_UTILS_MODULE_URL,
                    runtime: impFlattenNestedViewRenderNodes
                });
                Identifiers.devModeEqual = new compile_metadata_1.CompileIdentifierMetadata({ name: 'devModeEqual', moduleUrl: CD_MODULE_URL, runtime: impDevModeEqual });
                Identifiers.interpolate = new compile_metadata_1.CompileIdentifierMetadata({ name: 'interpolate', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: impInterpolate });
                Identifiers.castByValue = new compile_metadata_1.CompileIdentifierMetadata({ name: 'castByValue', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: impCastByValue });
                Identifiers.pureProxies = [
                    null,
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy1', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy1 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy2', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy2 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy3', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy3 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy4', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy4 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy5', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy5 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy6', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy6 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy7', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy7 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy8', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy8 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy9', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy9 }),
                    new compile_metadata_1.CompileIdentifierMetadata({ name: 'pureProxy10', moduleUrl: VIEW_UTILS_MODULE_URL, runtime: view_utils_1.pureProxy10 }),
                ];
                return Identifiers;
            }());
            exports_1("Identifiers", Identifiers);
        }
    }
});
