// angular2/ts/src/compiler/output/path_util.js
System.register(['angular2/src/facade/exceptions', 'angular2/src/facade/lang'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var exceptions_1, lang_1;
    var _ASSET_URL_RE, _PATH_SEP, _PATH_SEP_RE, ImportEnv, _AssetUrl;
    /**
     * Returns the module path to use for an import.
     */
    function getImportModulePath(moduleUrlStr, importedUrlStr, importEnv) {
        var absolutePathPrefix = importEnv === ImportEnv.Dart ? "package:" : '';
        var moduleUrl = _AssetUrl.parse(moduleUrlStr, false);
        var importedUrl = _AssetUrl.parse(importedUrlStr, true);
        if (lang_1.isBlank(importedUrl)) {
            return importedUrlStr;
        }
        // Try to create a relative path first
        if (moduleUrl.firstLevelDir == importedUrl.firstLevelDir &&
            moduleUrl.packageName == importedUrl.packageName) {
            return getRelativePath(moduleUrl.modulePath, importedUrl.modulePath, importEnv);
        }
        else if (importedUrl.firstLevelDir == 'lib') {
            return "" + absolutePathPrefix + importedUrl.packageName + "/" + importedUrl.modulePath;
        }
        throw new exceptions_1.BaseException("Can't import url " + importedUrlStr + " from " + moduleUrlStr);
    }
    exports_1("getImportModulePath", getImportModulePath);
    function getRelativePath(modulePath, importedPath, importEnv) {
        var moduleParts = modulePath.split(_PATH_SEP_RE);
        var importedParts = importedPath.split(_PATH_SEP_RE);
        var longestPrefix = getLongestPathSegmentPrefix(moduleParts, importedParts);
        var resultParts = [];
        var goParentCount = moduleParts.length - 1 - longestPrefix;
        for (var i = 0; i < goParentCount; i++) {
            resultParts.push('..');
        }
        if (goParentCount <= 0 && importEnv === ImportEnv.JS) {
            resultParts.push('.');
        }
        for (var i = longestPrefix; i < importedParts.length; i++) {
            resultParts.push(importedParts[i]);
        }
        return resultParts.join(_PATH_SEP);
    }
    exports_1("getRelativePath", getRelativePath);
    function getLongestPathSegmentPrefix(arr1, arr2) {
        var prefixSize = 0;
        var minLen = lang_1.Math.min(arr1.length, arr2.length);
        while (prefixSize < minLen && arr1[prefixSize] == arr2[prefixSize]) {
            prefixSize++;
        }
        return prefixSize;
    }
    exports_1("getLongestPathSegmentPrefix", getLongestPathSegmentPrefix);
    return {
        setters:[
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            // asset:<package-name>/<realm>/<path-to-module>
            _ASSET_URL_RE = /asset:([^\/]+)\/([^\/]+)\/(.+)/g;
            _PATH_SEP = '/';
            _PATH_SEP_RE = /\//g;
            (function (ImportEnv) {
                ImportEnv[ImportEnv["Dart"] = 0] = "Dart";
                ImportEnv[ImportEnv["JS"] = 1] = "JS";
            })(ImportEnv || (ImportEnv = {}));
            exports_1("ImportEnv", ImportEnv);
            _AssetUrl = (function () {
                function _AssetUrl(packageName, firstLevelDir, modulePath) {
                    this.packageName = packageName;
                    this.firstLevelDir = firstLevelDir;
                    this.modulePath = modulePath;
                }
                _AssetUrl.parse = function (url, allowNonMatching) {
                    var match = lang_1.RegExpWrapper.firstMatch(_ASSET_URL_RE, url);
                    if (lang_1.isPresent(match)) {
                        return new _AssetUrl(match[1], match[2], match[3]);
                    }
                    if (allowNonMatching) {
                        return null;
                    }
                    throw new exceptions_1.BaseException("Url " + url + " is not a valid asset: url");
                };
                return _AssetUrl;
            }());
        }
    }
});
